### DISCUSSION TITLE
Initial config in Docker

### DISCUSSION THREAD

**Discussion by zswickliffe** [juftin/camply#96] [Q&A] [answered]
Hey there, I'm able to get things up and running in docker, but in what seems like a round-about way and I wanted to see if there's a better way. 

For example, if I try to start a container by using: 
`docker run -d -i --name PNW_Sites juftin/camply camply campsites --rec-area 2835 --rec-area 2881 --rec-area 1120 --start-date 2022-07-30 --end-date 2022-08-01 --nights 2 --continuous --polling-interval 5 --search-forever --notifications email`

It will create the container and the container will run then immediately exit due to email notifications not being configured yet:
`2022-07-13 00:49:18,021 [   ERROR]: Email Notification Auth Parameters not set. Run `camply configure` or set the following Environment Variables:
        EMAIL_TO_ADDRESS
        EMAIL_USERNAME
        EMAIL_PASSWORD
Optional Environment Variables:
        EMAIL_SMTP_SERVER (default: smtp.gmail.com)
        EMAIL_FROM_ADDRESS (default: camply@juftin.com)
        EMAIL_SUBJECT_LINE (default: "Camply Notification")
        EMAIL_SMTP_PORT (default: 465)
2022-07-13 00:49:18,022 [  CAMPLY]: Exiting camply 👋`

So instead, I create a container by using this: 
`docker run -d -i --name PNW_Sites juftin/camply camply configure`

Which allows me to keep it running long enough to actually configure email notifications via terminal by typing `camply configure` and saving, then I have to open the terminal on the container and manually enter `camply campsites --rec-area 2835 --rec-area 2881 --rec-area 1120 --start-date 2022-07-30 --end-date 2022-08-01 --nights 2 --continuous --polling-interval 5 --search-forever --notifications email` in order to actually start the search correctly. 

I'm a total docker noob and self-taught so apologies if I'm totally missing something very obvious... My desire is to have it auto-search in the background when the container is started and avoid having to manually input via terminal each time it's loaded.

**Comment by juftin** [comment, marked as answer]
hey @zswickliffe this is a great question, I'm so happy to share some docker / camply knowledge. I've got some more information [here](https://juftin.com/camply/README.html#running-in-docker) so don't mind some of the duplicated content in my response.

When you run `camply configure` locally or in docker it creates a file called [`.camply` in your home folder](https://github.com/juftin/camply/blob/main/docs/examples/example.camply) (the home folder is in different default locations depending on whether you're on a mac, linux, or pc). Here's what that `.camply` file looks like: 
```
# ANY VARIABLES THAT ARE SET TO "" ARE REQUIRED, VARIABLES THAT HAVE A
    # VALUE PROVIDED ARE THE DEFAULT AND CAN BE CHANGED OR REMOVED ENTIRELY

# REQUIRED TO SEND PUSHOVER NOTIFICATIONS
PUSHOVER_PUSH_USER=""

# REQUIRED TO SEND PUSHBULLET NOTIFICATIONS
PUSHBULLET_API_TOKEN=""

# REQUIRED TO SEND TELEGRAM NOTIFICATIONS
TELEGRAM_BOT_TOKEN=""
TELEGRAM_CHAT_ID=""

# REQUIRED TO SEND EMAIL NOTIFICATIONS
EMAIL_TO_ADDRESS=""
EMAIL_USERNAME=""
EMAIL_PASSWORD=""
EMAIL_SMTP_SERVER="smtp.gmail.com"
EMAIL_SMTP_PORT=465
EMAIL_FROM_ADDRESS="camply@juftin.com"
EMAIL_SUBJECT_LINE="Camply Notification"

# PERSONAL PUSHOVER API TOKEN (OPTIONAL)
PUSHOVER_PUSH_TOKEN=""
# PERSONAL RECREATION.GOV API KEY: https://ridb.recreation.gov/profile (OPTIONAL)
RIDB_API_KEY=""
```

When camply detects a `.camply` file in the home folder it will automatically use it to set environment variables corresponding to the file (i.e. `EMAIL_TO_ADDRESS` on a `.camply` file becomes the `EMAIL_TO_ADDRESS` environment variable at runtime). If you don't have a `.camply` file you can also separately set the environment variables and camply will use them too. 

So for your docker container there are two ways to pass your email set up, either by setting the environment variables or by sharing a local `.camply` file into the docker container's home folder. Here's what those two options would look like for you:

```console
docker run -d \
  --name PNW_Sites \
  --env EMAIL_TO_ADDRESS="example@gmail.com" \
  --env EMAIL_USERNAME="example@gmail.com" \
  --env EMAIL_PASSWORD="password12345" \
  juftin/camply:latest \
  camply campsites \
    --rec-area 2835 \
    --rec-area 2881 \
    --rec-area 1120 \
    --start-date 2022-07-30 \
    --end-date 2022-08-01 \
    --nights 2 \
    --continuous \
    --polling-interval 5 \
    --search-forever \
    --notifications email
```

Or alternatively you can create your own `.camply` on your laptop (either with a text editor or by installing camply locally and running `camply configure`) and share it in to the docker container itself: 

```console
docker run -d \
  --name PNW_Sites \
  --volume /local/path/to/.camply:/home/camply/.camply \
  juftin/camply:latest \
  camply campsites \
    --rec-area 2835 \
    --rec-area 2881 \
    --rec-area 1120 \
    --start-date 2022-07-30 \
    --end-date 2022-08-01 \
    --nights 2 \
    --continuous \
    --polling-interval 5 \
    --search-forever \
    --notifications email
```

[Using a YAML Configuration file](https://juftin.com/camply/README.html#using-a-yml-configuration-file-to-search-for-campsites) is another nice way to do it in docker so I recommend checking that out too - you can share it into the docker container just like the `.camply` file. 

One other really nice way to do all of this is with docker compose where you can put all of the docker configuration into a `docker-compose` file.

Here's your `docker-compose.yaml`:
```yaml
services:
    camply:
        container_name: PNW_Sites
        image:          juftin/camply:latest
        environment:
            EMAIL_TO_ADDRESS: "example@gmail.com"
            EMAIL_USERNAME:   "example@gmail.com"
            EMAIL_PASSWORD:   "password12345"
        command:        >
                        camply campsites
                            --rec-area 2835
                            --rec-area 2881
                            --rec-area 1120
                            --start-date 2022-07-30
                            --end-date 2022-08-01
                            --nights 2
                            --continuous
                            --polling-interval 5
                            --search-forever
                            --notifications email
```

then you can bring it up with `docker compose up -d`.

Hopefully this all helps, let me know if you're able to get it going.

**Comment by zswickliffe** [comment]
Oh man, what an amazing response. I think I'm finally understanding the way docker works (previously I was just running in ubuntu). 

Thank you so much!!
